### PR TITLE
Add documentation for the public C++ API in the main Sphinx docs.

### PR DESCRIPTION
**Pull request by rmjarvis**
A long time ago, we used to use Doxygen for our docs, and had it make both C++ and python documentation.  It never looked very pretty, so we switched to Sphinx for the python.  But Sphinx doesn't really do C++, so we just dropped the C++ docs at the time.  

However, in response to issue #1150, I looked into it and discovered that there is a project called Breathe, which can take doxygen output and put it in Sphinx docs.  So I tried it out last night and it seems to work ok.  It's not super pretty, but it's basically functional.  

This PR adds a first pass at this, adding a C++ Layer section to the left side bar in the docs. I think I added all the relevant C++ functions and classes that we expect C++ GalSim users to potentially care about, skipping a lot of the implementation details.  I also didn't try to improve out doxygen markup at all, so whatever was there is all we have right now.  I also added some quick text giving an overview of how people can get the C++ library to link to and such.

I think this is probably fine for now as a straw man for people to improve later if they are so motivated.

### CODE REVIEW

File: CHANGELOG.rst [hunk 1]
```diff
@@ -24,6 +24,11 @@ Dependency Changes
 API Changes
 -----------
 
+- Removed CppEllipse in C++ layer, which had been deprecated since the 1.x series, but we forgot
+  to actually get rid of. (#1129)
+- Removed AstronomicalConstants.h in C++ layer, which we never used. (#1129)
+- Removed AttributeDict, which we had stting in utilities, but which we have never used.
```
Comment (rmjarvis) [inline CHANGELOG.rst:30]:
```suggestion
- Removed AttributeDict, which we had sitting in utilities, but which we have never used.
```

File: docs/cpp.rst [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+While GalSim is primarily a Python package, much of the implementation is done in C++
+for improved speed for the hard-core numerical calculations.
+
+If you would like to use the C++-layer functions in your own C++ project, you can do
+so with the caveat that these are officially implementation details, so we don't
+strictly enforce semantic versioning for the C++-layer API.  That is, function signatures
+may change on minor version updates (e.g. 2.3.x to 2.4.0).  We don't often make huge
+changes to the C++ API, so most of the time you will be fine upgrading, but you should
+be prepared that you may need to update your code when upgrading GalSim.  (We do
+guarantee that we won't change the C++ API for bugfix updates, e.g. 2.4.1 to 2.4.2.)
```
Comment (beckermr) [inline docs/cpp.rst:13]: API or ABI? 

In other words, can people who build their code against 2.4.1 link against 2.4.2 or do they need to recompile? If they need to recompile, then I think it is important to mention this.

Comment (rmjarvis) [reply docs/cpp.rst:13]: Probably mostly ABI, but I don't know really how to guarantee that.  I.e. I don't know what tests to do to ensure that we keep things ABI compatible.  There is nothing in our test suite that checks this.

Comment (rmjarvis) [reply docs/cpp.rst:13]: Ah, OK.  That I can do!  :)

**Review by beckermr — APPROVED** [review]

**Comment by rmjarvis** [comment]
Thanks Matt!

**Pull request merged**